add sum encoding and handling for sum types of unary constructors

justinwoo opened the pull request:
fixes #25 

uses some flags to determine casing and then uses these flags in encoding/decoding

updated excerpt:

```purs
testOption (makeCasingOptions toUpper) "\"FrIkAnDeL\"" Frikandel
testOption (makeCasingOptions toLower) "\"FrIkAnDeL\"" Frikandel
```

justinwoo commented:
Alternatively, I wrote this up where I've defined a function to provide instances specifically for sum types of unary constructors: https://github.com/justinwoo/purescript-foreign-generic/commit/179831cd3c21b084942c1dbb2a1383ead8f8afb7. Should I squash this here instead of using the conditional version in this branch?

I'd really like to also provide some `Fail` constrained instances for warnings and explanations, but I don't know how to do that without having overlapped instances warnings in usage sites.

paf31 commented:
What do you think about making a separate generic decoding/encoding function for enums like this? That way we could actually enforce using the type system that they only accept "enum types".

justinwoo commented:
I squashed my separate genericDecodeUnaryConstructors and genericEncodeUnaryConstructors functions into this one. Like my comment before though, while it doesn't match any instances for reps that aren't Sum or Constructor NoArg, I can't add Fail constraints without having overlapping instance warnings, so it could be a little confusing to use. But that fail constraint can wait until something with instance chains gets figured out.

paf31 commented on src/Data/Foreign/Generic/SumEncoding.purs:
```diff
@@ -0,0 +1,62 @@
+module SumEncoding where
```
Module name is incomplete here.

justinwoo replied:
oh, oops

paf31 commented on src/Data/Foreign/Generic/SumEncoding.purs:
```diff
@@ -0,0 +1,62 @@
+  => GenericDecodeUnarySum (Constructor name NoArguments) where
+  decodeSum {constructorTagTransform} f = do
+    tag <- constructorTagTransform <$> readString f
+    unless (constructorTagTransform tag == ctorName) $
```
You apply `constructorTagTransform` twice.

paf31 commented on src/Data/Foreign/Generic/SumEncoding.purs:
```diff
@@ -0,0 +1,62 @@
+      fail (ForeignError ("Expected " <> show ctorName <> " tag for unary constructor literal " <> ctorName))
+    pure $ Constructor NoArguments
+    where
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
```
And here

paf31 commented:
Just add a `Fail` instance for `Product` and `Rec`?

Also, can we use the term "enum"? I think it's pretty common for this type of data.

justinwoo commented:
Updated. I added the instances, though, I haven't been able to come up with ways to reach those without a constructor constraint.

paf31 commented:
Yes, you would need to write several instances for `Constructor`, but they won't overlap.

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,77 @@
+  encodeEnum _ _ = unsafeCrashWith "unreachable encodeEnum was reached."
+
+instance recGenericEncodeEnum
+  :: Fail """
```
Maybe pull the error messages out into a single common type synonym?

justinwoo replied:
But I wouldn't be able to change the wording right? I do want to put what kind of type they were trying to plug in here for a more relevant error message

paf31 replied:
Why not just say "doesn't work with constructors with arguments" though? Records and products are both types of arguments. Do you not think it's clear enough?

justinwoo replied:
I guess that's fine. I'll switch them over then

paf31 commented:
Another option is to use `TypeEquals`.

paf31 commented on src/Data/Foreign/Generic/Types.purs:
```diff
@@ -4,8 +4,11 @@ type Options =
   { sumEncoding :: SumEncoding
   , unwrapSingleConstructors :: Boolean
   , unwrapSingleArguments :: Boolean
+  , constructorTagTransform :: String -> String
```
Can this go into `TaggedObject`?

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,84 @@
+  decodeEnum opts f = Inl <$> decodeEnum opts f <|> Inr <$> decodeEnum opts f
+
+instance ctorNoArgsGenericDecodeEnum
+  :: (IsSymbol name)
```
Parens are redundant here.

justinwoo commented:
Updated this. Like we found on Slack, I couldn't find out how to make a synonym actually work at the usage site of the Fail instances, and then I moved the tag transform into SumEncoding.

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,105 @@
+
+instance ctorArgumentGenericEncodeEnum
+  :: Fail """
+    genericEncodeEnum cannot be used with constructors with arguments, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
```
I would still trim this down a bit and put the stuff about enums into the documentation for the classes.

justinwoo commented:
Wrote up some docstrings with a small code example in each top level method, with just short descriptions of the typeclasses and the instance errors.

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,126 @@
+  => GenericDecodeEnum (Constructor name NoArguments) where
+  decodeEnum {sumEncoding: TaggedObject {constructorTagTransform}} f = do
+    tag <- readString f
+    unless (constructorTagTransform tag == ctorName) $
```
Wrong way around here, surely?

justinwoo replied:
ah, yeah. the ctorname is already transformed, but i guess transforming the tag itself is kinda wonky...

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,126 @@
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
+
+instance ctorArgumentGenericDecodeEnum
+  :: Fail """
```
Please put these `Fail` constraints on one line and just use a single error message in a `type` declaration.

paf31 commented on src/Data/Foreign/Generic/Enum.purs:
```diff
@@ -0,0 +1,126 @@
+  :: forall a rep
+   . Generic a rep
+  => GenericDecodeEnum rep
+  => Options
```
If we're only using `constructorTagTransform` here, let's only require a record with that one field.

paf31 replied:
We can use row polymorphism if we want to support more fields.

paf31 commented:
Sorry for the delay, I have a few more comments.

justinwoo commented:
I updated this, but I still cannot get the type alias string to properly work on my machine for the Fail constraint. I will get "No type class instance was found for Fail GenericEncodeDecodeFailMessage"...

paf31 commented:
Ok, I'd prefer one of the following then:

1. No `Fail` instances for now
2. Just repeating the same error message in the code, but on single lines.
3. Adding a helper class with a single instance so that we can put the error in a single place.

justinwoo commented:
Pushed an update to do no 2. I'll try to remember to come back to this when we get https://github.com/purescript/purescript/issues/2972 fixed.

paf31 commented:
Great, thanks for all the work on this!

paf31 commented on bower.json:
```diff
@@ -16,15 +16,16 @@
-    "purescript-symbols": "^3.0.0"
+    "purescript-proxy": "^2.1.0",
+    "purescript-symbols": "^3.0.0",
+    "purescript-strings": "^3.2.1"
```
It doesn't look like you're using this any more.

justinwoo replied:
😨

justinwoo commented:
Thanks!!! Now I need to remember to update my howto guides and start making some noise about enum sum type support 😄
